src/poetry/puzzle/provider.py

```diff
@@ -945,19 +945,9 @@ def _merge_dependencies_by_constraint(
         by_constraint: dict[VersionConstraint, list[Dependency]] = defaultdict(list)
         for dep in dependencies:
             by_constraint[dep.constraint].append(dep)
-        for constraint, _deps in by_constraint.items():
-            new_markers = []
-            for dep in _deps:
-                marker = dep.marker.without_extras()
-                if marker.is_any():
-                    # No marker or only extras
-                    continue
-
-                new_markers.append(marker)
-
-            if not new_markers:
-                continue
 
+        for constraint, _deps in by_constraint.items():
+            new_markers = [d.marker for d in _deps]
```

Comment:
Does that maybe even fix #5429? I'd appreciate if you could check and if that's the case split this change into a separate PR and add a test. 🙏 

By the way, line 952 can probably be simplified to `dep.marker = MarkerUnion.of(*new_markers)`. If I don't miss something the marker of the first dependency is considered twice unnecessarily at the moment.

Reply:
I considered MarkerUnion.of() but it's heading in the wrong direction with respect to https://github.com/python-poetry/poetry-core/pull/530, which removes much of the simplification from the `of()` methods

Reply:
From an API perspective, I wonder if it might make sense to have two versions of `of()` (better names welcome). One that we use internally for cnf/dnf and one (that basically just does the `dnf(cnf())` trick) that can be used in cases like this. It seems more natural to use something like `of()` here. Or maybe just add a parameter to `of()` which is actively set by cnf/dnf?

Reply:
so far as I can see this doesn't touch #5429

https://github.com/python-poetry/poetry-core/pull/530 is more-or-less making `XXXMarker.of()` a private method.  

I'm going to tweak this particular line so that it goes `AnyMarker().union(MarkerUnion(*new_markers)` which is certainly a bit artificial.  But I think it's the most efficient thing to do, or anyway will be if-and-when https://github.com/python-poetry/poetry-core/pull/530 merges.

I wouldn't be opposed to some new API allowing union-of-lots-of-markers to be more conveniently constructed in one go, but also I'm not especially excited about it.

Reply:
I took a closer look. It's part of the issue. See #7175

Reply:
Oh, how satisfying.  Nice.